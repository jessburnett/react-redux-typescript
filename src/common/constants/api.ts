const apiBaseUrl = '/api';
export const dataApiBaseUrl = `${apiBaseUrl}/get-data`;
