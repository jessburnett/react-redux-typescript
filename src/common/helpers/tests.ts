export const testGeneratorStep = (generator, step: number = 1, datas = {}) => {
  let stepResult;
  for (let i = 0; i < step; i++) {
    stepResult = datas[i + 1] ? generator.next(datas[i + 1]) : generator.next();
  }
  return stepResult;
};
