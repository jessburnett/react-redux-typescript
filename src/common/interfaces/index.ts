import { IAction, IEmptyAction, IFailureAction } from './IAction';
import { IError } from './IError';
import { IImmutableArray } from './IImmutableArray';
import { IImmutableObject } from './IImmutableObject';

export {
  IAction,
  IEmptyAction,
  IFailureAction,
  IError,
  IImmutableArray,
  IImmutableObject,
};
