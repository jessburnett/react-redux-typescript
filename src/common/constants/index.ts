import * as actions from './actions';
import { dataApiBaseUrl } from './api';

export {
  actions,
  dataApiBaseUrl,
};
