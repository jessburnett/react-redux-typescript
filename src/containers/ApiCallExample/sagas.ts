import { takeLatest, put } from 'redux-saga/effects';

import { GET_DATA } from 'common/constants/actions';
import { getDataSuccess, getDataFailure } from 'store/actions';

import { fetchData } from './api';

export function* getData() {
  try {
    const data = yield fetchData();
    yield put(getDataSuccess(data));
  } catch (error) {
    yield put(getDataFailure(error));
  }
}

export default function* watch() {
  yield takeLatest(GET_DATA, getData);
}
