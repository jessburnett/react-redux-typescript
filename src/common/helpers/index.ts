import { testGeneratorStep } from './tests';

export {
  testGeneratorStep,
};
