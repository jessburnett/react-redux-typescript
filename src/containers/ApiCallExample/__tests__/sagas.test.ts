import { actions } from 'common/constants';
import { testGeneratorStep } from 'common/helpers';

const fetchDataMock = jest.fn();
jest.mock('../api', () => ({
  fetchData: fetchDataMock,
}));

import { getData } from '../sagas';

const testData = [1,2,3];

describe('ApiCallExample sagas', () => {

  describe('get data', () => {
    let generator;

    beforeEach(() => {
      generator = getData();
      fetchDataMock.mockClear();
    });

    it('fetches data properly', () => {
      fetchDataMock.mockImplementationOnce(() => testData);
      const { value } = testGeneratorStep(generator);
      expect(value).toEqual(testData);
    });

    it(`dispatches ${actions.GET_DATA_SUCCESS} action properly`, () => {
      expect(testGeneratorStep(generator, 2, { 2: { data: testData }})).toMatchSnapshot();
    });

    it(`dispatches ${actions.GET_DATA_FAILURE} action properly`, () => {
      fetchDataMock.mockImplementationOnce(() => { throw 'error'; });
      expect(testGeneratorStep(generator, 2, { 2: { error: true }})).toMatchSnapshot();
    });

  });

});
