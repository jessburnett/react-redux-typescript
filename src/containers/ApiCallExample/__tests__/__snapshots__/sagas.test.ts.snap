exports[`ApiCallExample sagas get data dispatches GET_DATA_FAILURE action properly 1`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`ApiCallExample sagas get data dispatches GET_DATA_SUCCESS action properly 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": Object {
          "data": Object {
            "data": Array [
              1,
              2,
              3,
            ],
          },
        },
        "type": "GET_DATA_SUCCESS",
      },
      "channel": null,
    },
  },
}
`;
